File: date-picker/src/main/java/me/foolishchow/android/datepicker/options/DatePickerOption.java
package me.foolishchow.android.datepicker.options;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Calendar;
import java.util.Date;

import me.foolishchow.android.datepicker.OnDatePickerSelectListener;
import me.foolishchow.android.datepicker.Utils;
import me.foolishchow.android.datepicker.WheelTime;

/**
 * Description:
 * Author: foolishchow
 * Date: 18/12/2020 3:42 PM
 */
public class DatePickerOption {

    @NonNull
    protected Calendar mRangeStart = sMinRangeStart;
    @NonNull
    protected Calendar mRangeEnd = sMaxRangeEnd;

    @WheelTime.DateStyle
    protected int mStyle = WheelTime.STYLE_DATE;

    @NonNull
    protected Calendar mSelected = Calendar.getInstance();

    @Nullable
    protected OnDatePickerSelectListener mOnDatePickerChangeListener;

    @NonNull
    public Calendar getRangeStart() {
        return mRangeStart;
    }

    public void setRangeStart(@Nullable Date rangeStart) {
        setRangeStart(rangeStart == null ? null : Utils.asCalendar(rangeStart));
    }

    public void setRangeStart(int year, int month, int dayOfMonth,
                              int hourOfDay, int minute, int second) {
        setRangeStart(Utils.asCalendar(year, month, dayOfMonth, hourOfDay, minute, second));
    }

    public void setRangeStart(int year, int month, int dayOfMonth) {
        setRangeStart(Utils.asCalendar(year, month, dayOfMonth, 0, 0, 0));
    }

    public void setRangeStart(@Nullable Calendar rangeStart) {
        if (rangeStart == null) {
            rangeStart = sMinRangeStart;
        } else if (rangeStart.getTimeInMillis() < sMinRangeStart.getTimeInMillis()) {
            rangeStart = sMinRangeStart;
        }
        if (mRangeEnd.getTimeInMillis() < rangeStart.getTimeInMillis()) {
            throw new IllegalStateException("range start is later than range end !");
        }
        mRangeStart = rangeStart;
    }


    @NonNull
    public Calendar getRangeEnd() {
        return mRangeEnd;
    }

    public void setRangeEnd(int year, int month, int dayOfMonth) {
        setRangeEnd(Utils.asCalendar(year, month, dayOfMonth, 0, 0, 0));
    }

    public void setRangeEnd(int year, int month, int dayOfMonth,
                            int hourOfDay, int minute, int second
    ) {
        setRangeEnd(Utils.asCalendar(year, month, dayOfMonth, hourOfDay, minute, second));
    }


    public void setRangeEnd(@Nullable Date rangeEnd) {
        setRangeEnd(rangeEnd == null ? null : Utils.asCalendar(rangeEnd));
    }

    public void setRangeEnd(@Nullable Calendar rangeEnd) {
        if (rangeEnd == null) {
            rangeEnd = sMaxRangeEnd;
        } else if (rangeEnd.getTimeInMillis() > sMaxRangeEnd.getTimeInMillis()) {
            rangeEnd = sMaxRangeEnd;
        }
        if (mRangeStart.getTimeInMillis() > rangeEnd.getTimeInMillis()) {
            throw new IllegalStateException("range start is later than range end !");
        }
        mRangeEnd = rangeEnd;
    }

    @NonNull
    public Calendar getSelected() {
        return mSelected;
    }


    public void setSelected(int year, int month, int dayOfMonth) {
        setSelected(Utils.asCalendar(year, month, dayOfMonth, 0, 0, 0));
    }

    public void setSelected(int year, int month, int dayOfMonth,
                            int hourOfDay, int minute, int second) {
        setSelected(Utils.asCalendar(year, month, dayOfMonth, hourOfDay, minute, second));
    }

    public void setSelected(@Nullable Date selected) {
        mSelected = selected == null ? Calendar.getInstance() : Utils.asCalendar(selected);
    }

    public void setSelected(@Nullable Calendar selected) {
        mSelected = selected == null ? Calendar.getInstance() : selected;
    }

    @WheelTime.DateStyle
    public int getStyle() {
        return mStyle;
    }

    public DatePickerOption setStyle(@WheelTime.DateStyle int style) {
        mStyle = style;
        return this;
    }


    @NonNull
    public OnDatePickerSelectListener getOnDatePickerChangeListener() {
        return mOnDatePickerChangeListener;
    }

    public DatePickerOption setOnDatePickerSelectListener(@NonNull OnDatePickerSelectListener listener) {
        mOnDatePickerChangeListener = listener;
        return this;
    }


    private static final Calendar sMinRangeStart = Utils.asCalendar(1900, 1, 1);
    private static final Calendar sMaxRangeEnd = Utils.asCalendar(2100, 12, 31);
}

File: date-picker/src/main/java/me/foolishchow/android/datepicker/options/DatePickerOption.kt
package me.foolishchow.android.datepicker.options

import me.foolishchow.android.datepicker.DateStyle
import me.foolishchow.android.datepicker.OnDatePickerSelectListener
import me.foolishchow.android.datepicker.Utils
import java.util.*

/**
 * Description:
 * Author: foolishchow
 * Date: 18/12/2020 3:42 PM
 */
@Suppress("UNUSED")
open class DatePickerOption {
    var rangeStart: Calendar = sMinRangeStart
        internal set
    var rangeEnd: Calendar = sMaxRangeEnd
        internal set

    @get:DateStyle.Style
    @DateStyle.Style
    var style = DateStyle.STYLE_DATE
        internal set
    var selected: Calendar = Calendar.getInstance()
        internal set
    private var mOnDatePickerChangeListener: OnDatePickerSelectListener? = null


    @JvmOverloads
    open fun setRangeStart(year: Int, month: Int, dayOfMonth: Int,
                           hourOfDay: Int = 0, minute: Int = 0, second: Int = 0) {
        setRangeStart(Utils.asCalendar(year, month, dayOfMonth, hourOfDay, minute, second))
    }

    open fun setRangeStart(rangeStart: Date?) {
        setRangeStart(if (rangeStart == null) null else Utils.asCalendar(rangeStart))
    }

    open fun setRangeStart(start: Calendar?) {
        val rangeStart = when {
            start == null -> {
                sMinRangeStart
            }
            start.timeInMillis < sMinRangeStart.timeInMillis -> {
                sMinRangeStart
            }
            else -> {
                start
            }
        }
        check(rangeEnd.timeInMillis >= rangeStart.timeInMillis) { "range start is later than range end !" }
        this.rangeStart = rangeStart
    }

    @JvmOverloads
    open fun setRangeEnd(year: Int, month: Int, dayOfMonth: Int,
                         hourOfDay: Int = 0, minute: Int = 0, second: Int = 0
    ) {
        setRangeEnd(Utils.asCalendar(year, month, dayOfMonth, hourOfDay, minute, second))
    }

    open fun setRangeEnd(rangeEnd: Date?) {
        setRangeEnd(if (rangeEnd == null) null else Utils.asCalendar(rangeEnd))
    }

    open fun setRangeEnd(end: Calendar?) {
        val rangeEnd = when {
            end == null -> {
                sMaxRangeEnd
            }
            end.timeInMillis > sMaxRangeEnd.timeInMillis -> {
                sMaxRangeEnd
            }
            else -> {
                end
            }
        }
        check(rangeStart.timeInMillis <= rangeEnd.timeInMillis) { "range start is later than range end !" }
        this.rangeEnd = rangeEnd
    }

    @JvmOverloads
    fun setSelected(year: Int, month: Int, dayOfMonth: Int,
                    hourOfDay: Int = 0, minute: Int = 0, second: Int = 0) {
        setSelected(Utils.asCalendar(year, month, dayOfMonth, hourOfDay, minute, second))
    }

    open fun setSelected(selected: Date?) {
        this.selected = if (selected == null) Calendar.getInstance() else Utils.asCalendar(selected)
    }

    open fun setSelected(selected: Calendar?) {
        this.selected = selected ?: Calendar.getInstance()
    }

    fun setStyle(@DateStyle.Style style: Int): DatePickerOption {
        this.style = style
        return this
    }

    val onDatePickerChangeListener: OnDatePickerSelectListener
        get() = mOnDatePickerChangeListener!!

    fun setOnDatePickerSelectListener(listener: OnDatePickerSelectListener): DatePickerOption {
        mOnDatePickerChangeListener = listener
        return this
    }

    companion object {
        private val sMinRangeStart = Utils.asCalendar(1900, 1, 1)
        private val sMaxRangeEnd = Utils.asCalendar(2100, 12, 31)
    }
}